File: lib/theming/palette_snapshot.dart
import 'package:flutter/material.dart';
import 'package:libmonet/libmonet.dart';

// Snapshot of [Palette]: captures token outputs at construction time and
// returns the same values thereafter.
class PaletteSnapshot extends Palette {
  // Background
  final Color _background;
  final Color _backgroundText;
  final Color _backgroundFill;
  final Color _backgroundBorder;
  final Color _backgroundHovered;
  final Color _backgroundSplashed;
  final Color _backgroundHoveredFill;
  final Color _backgroundSplashedFill;
  final Color _backgroundHoveredText;
  final Color _backgroundSplashedText;
  final Color _backgroundHoveredBorder;
  final Color _backgroundSplashedBorder;

  // Fill
  final Color _fill;
  final Color _fillBorder;
  final Color _fillHovered;
  final Color _fillSplashed;
  final Color _fillText;
  final Color _fillHoveredText;
  final Color _fillSplashedText;
  final Color _fillIcon;
  final Color _fillHoveredBorder;
  final Color _fillSplashedBorder;

  // Color
  final Color _color;
  final Color _colorText;
  final Color _colorIcon;
  final Color _colorBorder;
  final Color _colorHovered;
  final Color _colorHoveredText;
  final Color _colorHoveredBorder;
  final Color _colorSplashed;
  final Color _colorSplashedText;
  final Color _colorSplashedBorder;

  // Text
  final Color _text;
  final Color _textHovered;
  final Color _textHoveredText;
  final Color _textSplashed;
  final Color _textSplashedText;

  PaletteSnapshot._({
    required super.baseColor,
    required super.baseBackground,
    required Color background,
    required Color backgroundText,
    required Color backgroundFill,
    required Color backgroundBorder,
    required Color backgroundHovered,
    required Color backgroundSplashed,
    required Color backgroundHoveredFill,
    required Color backgroundSplashedFill,
    required Color backgroundHoveredText,
    required Color backgroundSplashedText,
    required Color backgroundHoveredBorder,
    required Color backgroundSplashedBorder,
    required Color fill,
    required Color fillBorder,
    required Color fillHovered,
    required Color fillSplashed,
    required Color fillText,
    required Color fillHoveredText,
    required Color fillSplashedText,
    required Color fillIcon,
    required Color fillHoveredBorder,
    required Color fillSplashedBorder,
    required Color color,
    required Color colorText,
    required Color colorIcon,
    required Color colorBorder,
    required Color colorHovered,
    required Color colorHoveredText,
    required Color colorHoveredBorder,
    required Color colorSplashed,
    required Color colorSplashedText,
    required Color colorSplashedBorder,
    required Color text,
    required Color textHovered,
    required Color textHoveredText,
    required Color textSplashed,
    required Color textSplashedText,
  })  : _background = background,
        _backgroundText = backgroundText,
        _backgroundFill = backgroundFill,
        _backgroundBorder = backgroundBorder,
        _backgroundHovered = backgroundHovered,
        _backgroundSplashed = backgroundSplashed,
        _backgroundHoveredFill = backgroundHoveredFill,
        _backgroundSplashedFill = backgroundSplashedFill,
        _backgroundHoveredText = backgroundHoveredText,
        _backgroundSplashedText = backgroundSplashedText,
        _backgroundHoveredBorder = backgroundHoveredBorder,
        _backgroundSplashedBorder = backgroundSplashedBorder,
        _fill = fill,
        _fillBorder = fillBorder,
        _fillHovered = fillHovered,
        _fillSplashed = fillSplashed,
        _fillText = fillText,
        _fillHoveredText = fillHoveredText,
        _fillSplashedText = fillSplashedText,
        _fillIcon = fillIcon,
        _fillHoveredBorder = fillHoveredBorder,
        _fillSplashedBorder = fillSplashedBorder,
        _color = color,
        _colorText = colorText,
        _colorIcon = colorIcon,
        _colorBorder = colorBorder,
        _colorHovered = colorHovered,
        _colorHoveredText = colorHoveredText,
        _colorHoveredBorder = colorHoveredBorder,
        _colorSplashed = colorSplashed,
        _colorSplashedText = colorSplashedText,
        _colorSplashedBorder = colorSplashedBorder,
        _text = text,
        _textHovered = textHovered,
        _textHoveredText = textHoveredText,
        _textSplashed = textSplashed,
        _textSplashedText = textSplashedText,
        // ignore: invalid_use_of_visible_for_testing_member
        super.base(
          contrast: 0.5,
          algo: Algo.apca,
        );

  factory PaletteSnapshot.capture(Palette s) {
    return PaletteSnapshot._(
      baseColor: s.color,
      baseBackground: s.background,
      background: s.background,
      backgroundText: s.backgroundText,
      backgroundFill: s.backgroundFill,
      backgroundBorder: s.backgroundBorder,
      backgroundHovered: s.backgroundHovered,
      backgroundSplashed: s.backgroundSplashed,
      backgroundHoveredFill: s.backgroundHoveredFill,
      backgroundSplashedFill: s.backgroundSplashedFill,
      backgroundHoveredText: s.backgroundHoveredText,
      backgroundSplashedText: s.backgroundSplashedText,
      backgroundHoveredBorder: s.backgroundHoveredBorder,
      backgroundSplashedBorder: s.backgroundSplashedBorder,
      fill: s.fill,
      fillBorder: s.fillBorder,
      fillHovered: s.fillHovered,
      fillSplashed: s.fillSplashed,
      fillText: s.fillText,
      fillHoveredText: s.fillHoveredText,
      fillSplashedText: s.fillSplashedText,
      fillIcon: s.fillIcon,
      fillHoveredBorder: s.fillHoveredBorder,
      fillSplashedBorder: s.fillSplashedBorder,
      color: s.color,
      colorText: s.colorText,
      colorIcon: s.colorIcon,
      colorBorder: s.colorBorder,
      colorHovered: s.colorHovered,
      colorHoveredText: s.colorHoveredText,
      colorHoveredBorder: s.colorHoveredBorder,
      colorSplashed: s.colorSplashed,
      colorSplashedText: s.colorSplashedText,
      colorSplashedBorder: s.colorSplashedBorder,
      text: s.text,
      textHovered: s.textHovered,
      textHoveredText: s.textHoveredText,
      textSplashed: s.textSplashed,
      textSplashedText: s.textSplashedText,
    );
  }

  // Background
  @override
  Color get background => _background;
  @override
  Color get backgroundText => _backgroundText;
  @override
  Color get backgroundFill => _backgroundFill;
  @override
  Color get backgroundBorder => _backgroundBorder;
  @override
  Color get backgroundHovered => _backgroundHovered;
  @override
  Color get backgroundSplashed => _backgroundSplashed;
  @override
  Color get backgroundHoveredFill => _backgroundHoveredFill;
  @override
  Color get backgroundSplashedFill => _backgroundSplashedFill;
  @override
  Color get backgroundHoveredText => _backgroundHoveredText;
  @override
  Color get backgroundSplashedText => _backgroundSplashedText;
  @override
  Color get backgroundHoveredBorder => _backgroundHoveredBorder;
  @override
  Color get backgroundSplashedBorder => _backgroundSplashedBorder;

  // Fill
  @override
  Color get fill => _fill;
  @override
  Color get fillBorder => _fillBorder;
  @override
  Color get fillHovered => _fillHovered;
  @override
  Color get fillSplashed => _fillSplashed;
  @override
  Color get fillText => _fillText;
  @override
  Color get fillHoveredText => _fillHoveredText;
  @override
  Color get fillSplashedText => _fillSplashedText;
  @override
  Color get fillIcon => _fillIcon;
  @override
  Color get fillHoveredBorder => _fillHoveredBorder;
  @override
  Color get fillSplashedBorder => _fillSplashedBorder;

  // Color (ink)
  @override
  Color get color => _color;
  @override
  Color get colorText => _colorText;
  @override
  Color get colorIcon => _colorIcon;
  @override
  Color get colorBorder => _colorBorder;
  @override
  Color get colorHovered => _colorHovered;
  @override
  Color get colorHoveredText => _colorHoveredText;
  @override
  Color get colorHoveredBorder => _colorHoveredBorder;
  @override
  Color get colorSplashed => _colorSplashed;
  @override
  Color get colorSplashedText => _colorSplashedText;
  @override
  Color get colorSplashedBorder => _colorSplashedBorder;

  // Text standalone
  @override
  Color get text => _text;
  @override
  Color get textHovered => _textHovered;
  @override
  Color get textHoveredText => _textHoveredText;
  @override
  Color get textSplashed => _textSplashed;
  @override
  Color get textSplashedText => _textSplashedText;
}

File: js/src/palette-snapshot.ts
import {type Argb, argbFromHex, hexFromArgb} from './color.js';
import {type Palette, type PaletteRole, paletteRoles} from './palette.js';

export type PaletteSnapshotRecord = Record<PaletteRole, Argb>;
export type PaletteSnapshotHexRecord = Record<PaletteRole, string>;

/**
 * Materialized snapshot of a Palette's role outputs.
 *
 * Dart's PaletteSnapshot captures every lazily-computed Palette role at a point
 * in time. The JS version mirrors that behavior with immutable ARGB role data
 * and the same role getters used by Palette.
 */
export class PaletteSnapshot {
  private readonly roles: PaletteSnapshotRecord;

  private constructor(record: PaletteSnapshotRecord) {
    this.roles = Object.freeze({...record}) as PaletteSnapshotRecord;
  }

  static capture(palette: Pick<Palette, PaletteRole | 'toRecord'>): PaletteSnapshot {
    return new PaletteSnapshot(palette.toRecord());
  }

  static fromRecord(record: Record<string, Argb>): PaletteSnapshot {
    const out = {} as PaletteSnapshotRecord;
    for (const role of paletteRoles) {
      const value = record[role];
      if (value === undefined) throw new Error(`Missing palette role: ${role}`);
      out[role] = value;
    }
    return new PaletteSnapshot(out);
  }

  static fromHexRecord(record: Record<string, string>): PaletteSnapshot {
    const out = {} as PaletteSnapshotRecord;
    for (const role of paletteRoles) {
      const value = record[role];
      if (value === undefined) throw new Error(`Missing palette role: ${role}`);
      out[role] = argbFromHex(value);
    }
    return new PaletteSnapshot(out);
  }

  private get(role: PaletteRole): Argb { return this.roles[role]; }

  get background(): Argb { return this.get('background'); }
  get backgroundText(): Argb { return this.get('backgroundText'); }
  get backgroundFill(): Argb { return this.get('backgroundFill'); }
  get backgroundBorder(): Argb { return this.get('backgroundBorder'); }
  get backgroundHovered(): Argb { return this.get('backgroundHovered'); }
  get backgroundSplashed(): Argb { return this.get('backgroundSplashed'); }
  get backgroundHoveredFill(): Argb { return this.get('backgroundHoveredFill'); }
  get backgroundSplashedFill(): Argb { return this.get('backgroundSplashedFill'); }
  get backgroundHoveredText(): Argb { return this.get('backgroundHoveredText'); }
  get backgroundSplashedText(): Argb { return this.get('backgroundSplashedText'); }
  get backgroundHoveredBorder(): Argb { return this.get('backgroundHoveredBorder'); }
  get backgroundSplashedBorder(): Argb { return this.get('backgroundSplashedBorder'); }
  get color(): Argb { return this.get('color'); }
  get colorText(): Argb { return this.get('colorText'); }
  get colorIcon(): Argb { return this.get('colorIcon'); }
  get colorHovered(): Argb { return this.get('colorHovered'); }
  get colorSplashed(): Argb { return this.get('colorSplashed'); }
  get colorHoveredText(): Argb { return this.get('colorHoveredText'); }
  get colorSplashedText(): Argb { return this.get('colorSplashedText'); }
  get colorHoveredIcon(): Argb { return this.get('colorHoveredIcon'); }
  get colorSplashedIcon(): Argb { return this.get('colorSplashedIcon'); }
  get colorBorder(): Argb { return this.get('colorBorder'); }
  get colorHoveredBorder(): Argb { return this.get('colorHoveredBorder'); }
  get colorSplashedBorder(): Argb { return this.get('colorSplashedBorder'); }
  get fill(): Argb { return this.get('fill'); }
  get fillText(): Argb { return this.get('fillText'); }
  get fillIcon(): Argb { return this.get('fillIcon'); }
  get fillHovered(): Argb { return this.get('fillHovered'); }
  get fillSplashed(): Argb { return this.get('fillSplashed'); }
  get fillHoveredText(): Argb { return this.get('fillHoveredText'); }
  get fillSplashedText(): Argb { return this.get('fillSplashedText'); }
  get fillHoveredIcon(): Argb { return this.get('fillHoveredIcon'); }
  get fillSplashedIcon(): Argb { return this.get('fillSplashedIcon'); }
  get fillBorder(): Argb { return this.get('fillBorder'); }
  get fillHoveredBorder(): Argb { return this.get('fillHoveredBorder'); }
  get fillSplashedBorder(): Argb { return this.get('fillSplashedBorder'); }
  get text(): Argb { return this.get('text'); }
  get textHovered(): Argb { return this.get('textHovered'); }
  get textSplashed(): Argb { return this.get('textSplashed'); }
  get textHoveredText(): Argb { return this.get('textHoveredText'); }
  get textSplashedText(): Argb { return this.get('textSplashedText'); }

  toRecord(): PaletteSnapshotRecord {
    return {...this.roles};
  }

  toHexRecord(): PaletteSnapshotHexRecord {
    const out = {} as PaletteSnapshotHexRecord;
    for (const role of paletteRoles) out[role] = hexFromArgb(this.roles[role]);
    return out;
  }

  equals(other: PaletteSnapshot): boolean {
    return paletteRoles.every(role => this.roles[role] === other.roles[role]);
  }
}
